src/types/mod.rs

```diff
@@ -774,3 +808,171 @@ impl<T: Object> ValueInternal for JsFunction<T> {
         unsafe { neon_runtime::tag::is_function(env.to_raw(), other.to_raw()) }
     }
 }
+
+/// A builder for making a JavaScript function call (e.g., `parseInt("42")`).
+///
+/// The builder methods make it convenient to assemble the call from parts:
+/// ```
+/// # use neon::prelude::*;
+/// # fn foo(mut cx: FunctionContext) -> JsResult<JsNumber> {
+/// # let global = cx.global();
+/// # let parseInt = global.get(&mut cx, "parseInt")?;
+/// # let parseInt: Handle<JsFunction> = parseInt.downcast_or_throw(&mut cx)?;
+/// let x: Handle<JsNumber> = parseInt
+///     .arg(cx.string("42"))
+///     .call(&mut cx)?;
+/// # Ok(x)
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct FunctionCall<'a> {
+    callee: Handle<'a, JsFunction>,
+    this: Handle<'a, JsValue>,
+    args: Vec<Handle<'a, JsValue>>,
+}
+
+/// A builder for making either a JavaScript function call (e.g., `parseInt("42")`)
+/// or constructor call (e.g., `new Array(16)`).
+///
+/// The builder methods make it convenient to assemble the call from parts:
+/// ```
+/// # use neon::prelude::*;
+/// # fn foo(mut cx: FunctionContext) -> JsResult<JsObject> {
+/// # let global = cx.global();
+/// # let url = global.get(&mut cx, "URL")?;
+/// # let url: Handle<JsFunction> = url.downcast_or_throw(&mut cx)?;
+/// let obj = url
+///     .arg(cx.string("https://neon-bindings.com"))
+///     .new(&mut cx)?;
+/// # Ok(obj)
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct Call<'a> {
+    callee: Handle<'a, JsFunction>,
+    args: Vec<Handle<'a, JsValue>>,
+}
+
+impl<'a> FunctionCall<'a> {
+    /// Set the value of `this` for the function call.
+    pub fn this<V: Value>(mut self, this: Handle<'a, V>) -> Self {
+        self.this = this.upcast();
+        self
+    }
+
+    /// Add an argument to the arguments list.
+    pub fn arg<V: Value>(mut self, arg: Handle<'a, V>) -> Self {
+        self.args.push(arg.upcast());
+        self
+    }
+
+    /// Add multiple arguments to the arguments list.
+    pub fn args<A: Arguments<'a>>(mut self, args: A) -> Self {
+        args.append(&mut self.args);
+        self
+    }
+
+    /// Make the function call. If the function returns without throwing, the result value
+    /// is downcast to the type `V`, throwing a `TypeError` if the downcast fails.
+    pub fn call<'b, C: Context<'b>, V: Value>(self, cx: &mut C) -> JsResult<'b, V> {
```

Comment:
It would be really nice to make this take `&self` since it doesn't need to consume `self` (for the Node-API call). In turn that would allow the other methods to take `&mut self` instead of `self` which makes for better optional chaining.

Reply:
Pushed a change that implements this!